### ISSUE TITLE
Allow subsequently loaded `ffxml` files to overwrite earlier types/parameters

### ISSUE THREAD

**Issue by jchodera** [pandegroup/openmm#1344]
A specific consequence of enabling mix-and-match `ffxml` files (#1298) is that we can run into situations where later `ffxml` files might need to overwrite earlier atom types and parameter definitions.

One case where it would be useful to allow this is for the current AMBER forcefields.  If we use AMBER typenames as OpenMM type and class names to allow support for small molecules, we run into the issue that allowing mix-and-match of different water or ion models requires @rafwiewiora dissect out these parameters and residue definitions from a single set of AMBER parameter files, which could be very complex.  Instead, it would be much easier if we could simply prepare separate parameter files for these alternative water and ion models and allow alternative choices to overwrite the earlier definitions if the user adds them to the end of the list of parameter sets.

Currently, attempting to redefine types and parameters throws an exception:
```
>>> from simtk.openmm.app import ForceField
>>> ff = ForceField('tip3p.xml', 'tip3p.xml')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/simtk/openmm/app/forcefield.py", line 117, in __init__
    self.loadFile(file)
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/simtk/openmm/app/forcefield.py", line 140, in loadFile
    self.registerAtomType(type.attrib)
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/simtk/openmm/app/forcefield.py", line 183, in registerAtomType
    raise ValueError('Found multiple definitions for atom type: '+name)
ValueError: Found multiple definitions for atom type: tip3p-O
```
Are there other ramifications for allowing this kind of atom type and parameter overwriting by later `ffxml` files?

**Comment by rafwiewiora** [comment]
Just to say a bit more about the particular kind of problems I was thinking about: say you're building a system that needs some neutralizing ions (Na+/Cl-) and uses a 'non-AMBER' (say multisite/dummy atom) metal model for e.g. Zn. 

Currently this is not a problem as the available ffxml files are simply missing Zn, while they provide Na+/Cl-. 

With the new design, we'll have an 'ions' ffxml, which passed to `ForceField` will read all the ions - in this case there must be a mechanism that replaces the loaded in 'AMBER' Zn parameters with the multisite/dummy atom ones.

**Comment by rafwiewiora** [comment]
With need to make a decision on a scheme for atom type naming too for metals and solvents (maybe this should go into a separate issue? or maybe not because it's pretty related). 

* Are we staying with the solvent naming scheme 'model-atom type' as in `tip3p-O`? 

* Are we going to include this kind of 'metadata' in atom type names for metals too or just keep them simple? If include the 'metadata' is it only model name / also oxidation state? For example: `Zn` vs `Zn(II)-Zn` vs `Zn-multiSite-Zn` vs `Zn(II)-multiSite-Zn` etc. 

Doing these complex atom type names for metals would alleviate the mix and matching problem to some extent (while I still think we should introduce mix&matching regardless to avoid future problems) - and it appears that it will be necessary in the future as we introduce separate oxidation states / ability for user to use multilple models of a particular metal element in the same system.

**Comment by jchodera** [comment]
> With need to make a decision on a scheme for atom type naming too for metals and solvents (maybe this should go into a separate issue?

Separate issue?

**Comment by peastman** [comment]
We need to be very careful about this, since it creates a lot of opportunities for errors.  For example, simply changing the order in which you list the xml files would change which ion model you get!  Also, while there are some situations where you really do intend one file to override another one, there are a lot of situations where you don't.  In that case, finding multiple definitions for the same types indicates a mistake, and we want to be able to catch it.

What about the following?

1. It will no longer matter what order you list the XML files in.  We will first load all atom types from all files, then all templates, then all forces.

2. The *only* thing that can be overridden is templates.  If a file wants to define a new Zn model, for example, it will provide its own atom types which are distinct from those in the main force field, along with any needed force parameters and bonds for them.

3. We will add a flag to the `<Residue>` tag indicating it is intended to override other templates defined elsewhere.  If two templates with the same name are found, the one containing the override flag is kept and the other is discarded.  If both have that flag, or neither has it, an exception is thrown.

**Comment by jchodera** [comment]
> For example, simply changing the order in which you list the xml files would change which ion model you get!

Note that this is actually the current behavior!  If you have `file1.xml` and `file2.xml` and both define atom types belonging to the same atom class, the order in which you specify the files controls which class-defined parameters get applied to which atom types.

**Comment by peastman** [comment]
No, the current situation is that if you list them in the wrong order, you get an exception that no parameters could be found for some atoms.  That's certainly not ideal, but it's not as bad as having both orders "work" but produce different results.

**Comment by jchodera** [comment]
Oh?
```
>>> from simtk.openmm.app import ForceField, PDBFile
>>> pdb = PDBFile('/Users/choderaj/miniconda/lib/python2.7/site-packages/simtk/openmm/app/data/tip3p.pdb')
>>> ff1 = ForceField('tip3p.xml', 'tip3pfb.xml')
>>> ff2 = ForceField('tip3pfb.xml', 'tip3p.xml')
>>> system1 = ff1.createSystem(pdb.topology)
>>> system2 = ff2.createSystem(pdb.topology)
>>> system1.getForce(2).getParticleParameters(0)
[Quantity(value=-0.834, unit=elementary charge), Quantity(value=0.3150752406575124, unit=nanometer), Quantity(value=0.635968, unit=kilojoule/mole)]
>>> system2.getForce(2).getParticleParameters(0)
[Quantity(value=-0.848448690103, unit=elementary charge), Quantity(value=0.317796456355, unit=nanometer), Quantity(value=0.652143528104, unit=kilojoule/mole)]
```

**Comment by peastman** [comment]
That looks like a bug!

**Comment by jchodera** [comment]
I think it's just the way the `class`es behave right now!  Parameters defined in terms of the `class` are only applied to the atom types that have been loaded *up to the point where the parameter is loaded*.  If another atom type is loaded after that, subsequent parameters defined in terms of `classes` apply to all previously loaded atom types in that class.  

I think this means we fundamentally need to change the way `class` is matched, or we need to defer expansion of atom classes into atom types.

**Comment by jchodera** [comment]
I think the above proposal (under "What about the following?") has merit, but we should think about the use cases.  For example, the AMBER forcefields define TIP3P water atom types (`HW`, `OW`) and residues, but we may want to switch those out with other water models.  We can be careful to make sure all water atom types and classes do not clash with these and specify them to work with your proposal above, but is there a simpler way atom type and parameter overwriting could work that could give us the behavior we want?

Use cases I envision:
* Using a different water model (probably most common)---may involve different parameters and the same or different residue template
* Using a different set of ion models (e.g. better parameters, multisite ions)---different parameters and same or different residue template
* Specifying some modified residues for prosthetic groups (e.g. heme) or ligands
* Maybe updated backbone parameters for a forcefield (though we may want to just do a whole forcefield conversion for updated forcefields)

**Comment by rafwiewiora** [comment]
I didn't hear this introduced to the discussion yet and it sounds like a good idea in my head that could solve some worries here - couldn't we have something like a `ff.overwrite('xyz.xml')` function that would be a nice explicit way of specifying any modifications / patches / etc. by the user? In this case `ForceField(xyz.xml, ...)` would continue to throw an exception if there were conflicting atom types or residues. The final picture of the `ff` would be determined by the order of `ff.overwrite` calls by the user.

**Comment by swails** [comment]
What happens right now if you load a `ffxml` file that re-defines an atom type, but gives it identical parameters?

In the Amber FF files, proteins and nucleic acids share the same sets of atom types.  Splitting these into separate XML files would require being able to define redundant types...

**Comment by rafwiewiora** [comment]
> What happens right now if you load a ffxml file that re-defines an atom type, but gives it identical parameters?

Exception, see @jchodera 's first post here for example. All that matters is that you're loading in an atom type with the same name twice. 

> In the Amber FF files, proteins and nucleic acids share the same sets of atom types. Splitting these into separate XML files would require being able to define redundant types...

Right, and hopefully we can make that problem go away by doing the templates - parameters separation, but it would be good to have some overwriting mechanism in place if some problems arise.

**Comment by peastman** [comment]
> couldn't we have something like a ff.overwrite('xyz.xml') function

That's also a possibility.  It's a bit less convenient for the user.  It does make it explicit what files you intend to take precedence though.

> Using a different water model (probably most common)---may involve different parameters and the same or different residue template

The method I suggested handles this very easily.  Each water model will define its own atom types.  There's no risk of collisions.  The only decision ForceField needs to make is which template to match to the residue.  Once that's decided, everything else is unambiguous.

> Using a different set of ion models (e.g. better parameters, multisite ions)---different parameters and same or different residue template

Same as above.

> Specifying some modified residues for prosthetic groups (e.g. heme) or ligands

That clearly requires a different template.  Depending on the modification, it might reuse some of the standard atom types, or define entirely new ones.

> Maybe updated backbone parameters for a forcefield (though we may want to just do a whole forcefield conversion for updated forcefields)

I would agree, that should be treated as a different force field.

**Comment by jchodera** [comment]
> That's also a possibility. It's a bit less convenient for the user. It does make it explicit what files you intend to take precedence though.

Listing the files in an order (`ForceField('file1.xml', 'file2.xml')`) is also an unambiguous, explicit way to specify precedence!

Another issue that has come up that bears on this is the difficulty we're having in carving up the CHARMM36 forcefield into independent pieces due to dependencies between files.  [This comment](https://github.com/ParmEd/ParmEd/issues/486#issuecomment-171822778) notes that many stream files have complex dependencies, so that we would seemingly need to have a lot of "bundled" `ffxml` files (that include multiple files), need to introduce some way to specify dependencies, or simply bundle the whole CHARMM36 forcefield into one big `ffxml` file with minimal carve-outs for different solvent model choices.  This last option is looking most likely for now, which makes it all the more necessary to have some way for users to override parts of this they need to modify.

> The method I suggested handles this very easily. Each water model will define its own atom types. There's no risk of collisions. The only decision ForceField needs to make is which template to match to the residue. Once that's decided, everything else is unambiguous.

Using a unique prefix for the atom types is sounding like a good idea for water and ion models.

Revisiting @peastman's suggestion:
> It will no longer matter what order you list the XML files in. We will first load all atom types from all files, then all templates, then all forces.
> The only thing that can be overridden is templates. If a file wants to define a new Zn model, for example, it will provide its own atom types which are distinct from those in the main force field, along with any needed force parameters and bonds for them.
> We will add a flag to the <Residue> tag indicating it is intended to override other templates defined elsewhere. If two templates with the same name are found, the one containing the override flag is kept and the other is discarded. If both have that flag, or neither has it, an exception is thrown.

I think this *could* work, but it sounds more complicated than it needs to be---especially the introduction of a new tag for `<Residue>` that allows it to override other templates. Is it possible to override the overrides? Do we have an integer precedence level that lets us specify the priority for each residue template over others, with a default of 0?  That could work, but is it really superior to "last one wins" for both residue templates and parameters?

The "last one wins" is used a great deal in specifying forcefield parameters like torsions, where general types match first and then more specific times specified later override them.

**Comment by peastman** [comment]
I'm thinking of it in terms of two questions: what are the important use cases, and what sorts of mistakes are people likely to make?

Listing the files in the wrong order is a really, really easy mistake.  It's inevitable that people will sometimes do it.  It also isn't obvious to users that the order *should* matter.  If I were just looking at a script, and saw it listing a set of XML files, I certainly wouldn't assume the order mattered.  So I'd prefer something where any overrides are much more explicit, where it's absolutely clear that someone (either the user or the force field designer) has thought about it and decided which one should take precedence.  It will lead to fewer mistakes.

Are there actually use cases that need multiple levels of overriding?  I can see the use cases where you would have two definitions for a residue (e.g. a file defining all your ions, and a separate file defining a multi-site model for one particular ion), but I don't know of any use cases where it would need to pick between three different definitions for a single residue.

**Comment by swails** [comment]
> The "last one wins" is used a great deal in specifying forcefield parameters like torsions, where general types match first and then more specific times specified later override them.

This isn't actually an illustrative example, as it's not "last one wins".  A specific torsion *always* overrides a generic torsion (i.e., one with wild-cards), regardless of which is specified "first" or "last".  The implementation is typically something along the lines of:

1. Look for a matching specific torsion
2. If one is not found, look through generic torsions and take any that match the middle two atom types (slightly different for impropers).

However, "last one wins" seems to be a fairly common convention.  I know Amber uses it, and I think CHARMM does, too.  No idea if GROMACS does or not.  But Amber primarily makes use of that feature through leaprc files, so it's the developers doing the overriding, not the users.

**Comment by jchodera** [comment]
> Are there actually use cases that need multiple levels of overriding? I can see the use cases where you would have two definitions for a residue (e.g. a file defining all your ions, and a separate file defining a multi-site model for one particular ion), but I don't know of any use cases where it would need to pick between three different definitions for a single residue.

These forcefields evolve in all manner of layered ways.  ff99sb was a set of modified (phi,psi) torsions on top of parm99 parameters, which in turn was an updated set of parameters that used the residue templates, ions, and solvents from parm94.  ff99sb-ildn then changed just *four* sets of (phi,psi) torsions on top of ff99sb.  Then there was ff99sb-ildnnmr that brought in an additional set of corrections from Best and Hummer based on NMR.

I worry that a single level of overriding is going to be limiting.  "Last one wins" is the simplest way to allow overrides, but I can also appreciate @peastman's concern, though I think this could be fixed with clear user documentation stressing this as a feature.  

If we do go the route of having a special attribute for overrides, how about having the attribute be specified at the `ffxml` file level, with an integer that defaults to 0:
```XML
<ForceField priority="3">
</ForceField>
```
Higher priority wins.  This way, you can have as many levels of priority as you want, and it is still independent of the order in which you specify things?

**Comment by peastman** [comment]
That's historically how force fields have evolved, but that doesn't mean it's the best way to distribute them.  As @swails pointed out, even in the case of Amber, overriding is mostly something done by developers, not users.  So yes, 99SB-ILDN is mostly identical to 99SB, but we still package it as a separate force field, and I think that's the right way to do it.  I would oppose a system that asked users to say, "I want 99SB, and then I want to apply the ILDN patch on top of it."  That's asking users to do work that really should be done by the force field developers.

**Comment by jchodera** [comment]
I think you are confusing the research field of molecular simulations with
an engineering field.

Amber was built in a way to enable exactly this kind of development. It was
possible to make small tweaks on top of tweaks. You don't need to be a
developer to write an frcmod file and add a line to leap to bring those
changes in.

We don't want to be *less* flexible than this, especially if there is no
benefit in doing so. Adding an integer priority isn't any more demanding
than adding an override flag. But asking users that need more than one
level of override to compile all of their changes into a flat ffxml file
and then override that *is* a big burden.

I am not even quite sure what you are disagreeing with at this point. The
idea of an integer priority? The concept of simulations as a research
science?

**Comment by peastman** [comment]
What I'm objecting to is asking regular users (people who run simulations but don't develop force fields) to combine files together in unnecessarily complicated ways.  A person should just be able to say, "I want 99SB-ILDN," and get it, without knowing or caring how that particular force field was developed, or what sequence of changes were applied to what earlier force fields.

I haven't expressed any opinion one way or the other about an integer priority.  I'm still trying to understand what the use cases are.  When I asked before about use cases, you described the history of how that force field was developed.  That isn't directly relevant to the question I asked about: how a user would run OpenMM.

**Comment by jchodera** [comment]
Parameterizing or tweaking forcefields as described above is a real life use case. There isn't really such a thing as "regular users".

We already went through an enumerated list of use cases above that you responded to. Are you looking for more use cases within a particular category of the above?

**Comment by peastman** [comment]
"Parameterizing force fields" is not a use case.  It's a broad category of activities.  I'm looking for *specific* use cases where, as part of doing that, it would be important for someone to have three or more definitions of a single residue present at once, and have ForceField select between them.

**Comment by jchodera** [comment]
I'm a bit confused now---are we specifically talking *only* about overriding residue template definitions, or are we also talking about overriding atom types and/or force parameters?

**Comment by peastman** [comment]
That was what I proposed:

> The only thing that can be overridden is templates. If a file wants to define a new Zn model, for example, it will provide its own atom types which are distinct from those in the main force field, along with any needed force parameters and bonds for them.

**Comment by rafwiewiora** [comment]
Some of my thoughts:

* @peastman 's proposal for overriding of templates with a flag in the Residue tag - very neat, I like this

* on the other hand, I think one should be careful not to only consider these generalized 'users' as the only group of people we want to make these changes for - first and foremost there reason we are proposing these changes is because we have found that they would make our work easier. But because this is how I think about this, I tend to strongly prefer solutions that will add a functionality that people who know about it can use, rather then change an existing functionality (i.e. add an `overwrite` function > change how order of xml's in a `ForceField` call behaves). 

So now to the first point, I think use cases are pretty well pointed out at this point. From my point of interest in metal ions - user needs to load in both `ions.xml` and `dummy_zinc.xml`, `dummy_zinc.xml` contains a special Zn residue, with an override flag - fantastic. 

To the second point: I make a dummy atom Zn file, it requires the CYM (negative CYS) residue so I put it in there, copied from an Amber ffxml. And then depending on whether an Amber ffxml I'm gonna use in my whatever further script has a CYM or not - `ForceField` works ok or gives me an exception if I have two files with the same Atom Types. So if I had an `.overwrite` option, that would save me a good load of anger. Or maybe @jchodera just wants to go from ff99SB to ff99SB-ildn to ff99SB-ildnnmr by overriding, instead of making separate ffxmls because he's bored on the weekend lol. You know, whatever, a few lines of code to be able to override parameters too and make any development / research work a bit easier - nice, the fact that I can't load in two ffxml files that contain even the same information, but there's some overlapping Atom Types - ridiculous, man wants to throw something at a screen kind of situation. You said yourself:

> As @swails pointed out, even in the case of Amber, overriding is mostly something done by developers, not users.

So why not make it a bit easier for developers to play around with that overriding? 

BUT I agree with @peastman this must be done in a way where person who doesn't know how / why this works cannot make a mistake. So if we go with the order of files in the `ForceField` call matters - I would say add a bool argument to `ForceField` to be able to turn this overriding on, otherwise - exception. If we add an `.overwrite` thing - problem solved too.

**Comment by rafwiewiora** [comment]
As far as the priority tags are concerned, I think the fact that @jchodera has to be proposing to be able to make this have various values means there is a need here to make this more versatile and play-aroundable than an 'average user' might ever need. And I think that level of complexity is unnecessary - just allow this to be doable in some more ways if it looks like the flag / priority flag way is too limiting.

**Comment by jchodera** [comment]
> The only thing that can be overridden is templates

I think we need the ability to override parameters too.

Use cases: 
* Suppose I want to tweak some torsion parameters for a couple of residues.  This is very common, and is probably the *most common* form of "forcefield improvement" we've seen.  Requiring a huge amount of infrastructure (writing whole complete `ffxml` files) to do this (when it is really unnecessary) would hinder use of OpenMM by making it significantly harder to do this.
* Suppose I want to tweak the GB parameters for one atom.  
* Suppose I want to fix the LJ radius of a carbonyl oxygen.  We've had to do this before after it was found that there was an issue with AMBER Lennard-Jones parameters with oxygen in hydration free energy calculations.
* Suppose I need to adjust the force constant for nitrile (CN) bonds for simulating Stark effect spectroscopy.

I have personally done all of these things before.  

We're currently re-evaluating how we tackle the CHARMM forcefield [in this thread](https://github.com/ParmEd/ParmEd/issues/561).  The CHARMM forcefield has a lot of stream (`.str`) files that have complex dependencies, sometimes depending on other stream files that have their own dependencies.  If we wanted it to "just work" if the user includes the corresponding `streamfile.xml` in OpenMM, we would need these to use the XML include mechanism:
```XML
<!ENTITY otherFile SYSTEM "otherFile.xml">
```
in order to automatically pull in its dependencies.  The difficulty here is that this might result in the same file being included multiple times, which would require some way for us to allow multiple *identical* definitions of atom types, parameters, and templates.  

It will likely be far easier for us to simply convert every file over individually and let the users fend for themselves in figuring out exactly which combinations of files are needed to make things work.  This isn't very friendly, but is likely just as *unfriendly* as CHARMM.  We can provide a bundled `charmm36.xml` that uses the XML include mechanism to bring in the files we *think* most users would need, and that may make things just friendly enough.

**Comment by peastman** [comment]
> Requiring a huge amount of infrastructure (writing whole complete ffxml files) to do this

You make a copy of the existing force field file, call it myModifiedForceField.xml or whatever, and then modify the parameters you want.  I wouldn't describe that as "a huge amount of infrastructure".  I recognize that isn't how CHARMM does it, and that's the context in which this subject came up.  What I'm trying to figure out though is whether CHARMM's approach is actually a better way.

When the entire force field is contained in a single file, it's really really easy to copy that file, modify whatever parts of it you want, and then know exactly what you have and where all your definitions are.  Not so when your force field is split between many files, some of which you wrote and some of which other people wrote, with complicated rules about which parts of which files override which other parts of which other files.  Yes, these are two different paradigms.  But I don't see how any of your examples is actually easier or takes you less time when using the latter approach rather than the former?

**Comment by jchodera** [comment]
> You make a copy of the existing force field file, call it myModifiedForceField.xml or whatever, and then modify the parameters you want. I wouldn't describe that as "a huge amount of infrastructure". I recognize that isn't how CHARMM does it, and that's the context in which this subject came up

This isn't how AMBER does it either.  Or gromacs from what I understand.

> When the entire force field is contained in a single file, it's really really easy to copy that file, modify whatever parts of it you want, and then know exactly what you have and where all your definitions are. 

This certainly isn't the way we recommend people write code; why should it be the way we recommend people tinker with forcefields?

It also isn't easy to recognize what has changed.  Yes, we could `diff` the files, but trivial changes that do not affect the forcefield processing could affect the diff...

**Comment by jchodera** [comment]
Let's see how things go with converting the CHARMM forcefield piecemeal.

Do you have a strong preference for the two options:
1. Provide a `charmm36.xml` that includes all the CHARMM `xml` files we think one would commonly want to use, and let the user figure out which `xml` file combinations to include if they want to do anything more complicated (like they have to do in CHARMM right now)
2. Provide a way to automatically handle dependencies so that including a CHARMM `xml` file would automatically bring in dependencies rather than just say atom types are undefined.  This would require some more sophisticated handling of dependencies or duplicate parameters that are *identical*.

**Comment by peastman** [comment]
Option 1 sounds safer.  That ensures a human has thought about which files should be combined.

**Comment by rafwiewiora** [comment]
from https://github.com/ParmEd/ParmEd/issues/486:

After separating an FF into: params, protein templates, nucleic templates separate ffxmls - it became clear to me that when calling `ForceField` you will have to pass the parameters file / files first, before the templates files. 

As such, this is really no longer an 'order doesn't matter' situation, and if parameters need to go first, I think that opens a way to also saying - overwriting happens with order. 

But aside from that, are we also looking for a solution to the 'what needs to comes first' problem, to which we could:

* have a more useful exception, right now for example `KeyError: 'N'`, there should be a suggestion that params need to be loaded first in this

* have `ForceField` `try` different combinations of the file order if there's an exception

* change from `ForceField` parsing file-by-file to merging all the xml tree's together and then parsing parameters first - then templates?

**Comment by peastman** [comment]
I'm still pretty negative about the idea of overriding based on order.  It's just way too easy for people to make mistakes, possibly without even realizing the order matters.  So I prefer something like your third option.  The order I would suggest is: 1) parse all atom types, 2) parse all templates, 3) parse all forces.

Then if we want overriding, it should be done in a way that's more explicit, and guarantees that someone (either the user or the force field developer) has thought about what should override what.  I'm open to suggestions for possible APIs.

**Comment by peastman** [comment]
Anyone want to offer suggestions on APIs?

Here are some options that occur to me:

- It's specified in the XML file.  This means overriding is the decision of the force field designer, not the user.  If that's possible it seems like a better option (since the force field designer is best qualified to know what should override what), but it may not always be possible.

- @rafwiewiora suggested having a method along the lines of `ff.override('something.xml')` that loads a new file, overriding previous definitions.

- We could do that as a constructor argument, something like `ff = ForceField('file1.xml', 'file2.xml', override=['file3.xml'])`.

- Or we could do it based on order, but you have to specify an argument to say, "Yes, I really want overriding to happen."  Something like `ff = ForceField('file1.xml', 'file2.xml', 'file3.xml', overriding=True)`.

Other suggestions?

We also still need to decide how overriding works.  Is it only templates that get overridden, and they use unique atom types?  Or is every generator supposed to individually support overriding and allow its parameter assignment rules to be overridden?

**Comment by jchodera** [comment]
These are good suggestions.

For the "specified in the XML file" option, I suggested an ffxml `ForceField` attribute:
```XML
<ForceField priority="3">
</ForceField>
```
Highest priority wins.  Default priority is 0 if `priority` is omitted.

My actual favorite is the `ff = ForceField('file1.xml', 'file2.xml', 'file3.xml', overriding=True)` option, but the ffxml option above is a close second.

**Comment by peastman** [comment]
The big difference between those two options is that the first one asks the force field developer to make the decision, while the second one asks the user to make the decision.  Could you comment on that?  Which of those is better?

**Comment by ChayaSt** [comment]
Here are some cases where CHARMM has more than one value for a parameter (There are more cases)
https://github.com/ParmEd/ParmEd/issues/561#issuecomment-173760099 

For cases like above, ```overriding=True``` is probably the better alternative because one is not necessarily higher priority than the other. And since CHARMM's approach is that the last one wins, we should do that too.

**Comment by rafwiewiora** [comment]
> My actual favorite is the ff = ForceField('file1.xml', 'file2.xml', 'file3.xml', overriding=True) option, but the ffxml option above is a close second.

Same here (i.e. I do like this more than my `.override` proposal). 

> The big difference between those two options is that the first one asks the force field developer to make the decision, while the second one asks the user to make the decision. Could you comment on that? Which of those is better?

Why don't we do both? We'll make sure FFs have the override tags for the residue that we are 100% clear must do the overriding, but also allow users to override themselves if need be. 

> We also still need to decide how overriding works. Is it only templates that get overridden, and they use unique atom types? Or is every generator supposed to individually support overriding and allow its parameter assignment rules to be overridden?

So I don't immediately see any use cases for 'standard uses' of the AMBER FFs - but I can imagine a power user / us doing some mixing of different parameter files and wanting overriding - @jchodera can you feed in more here on AMBER use cases? 

One thing re some metal stuff we discussed in the past I want to point out: using > 1 ion model for the same element - different oxidation states / different numbers of dummy atoms etc. - so perhaps we'll have two residues with different names and different atom types (say Fe2+ and Fe3+ with dummy atoms), they will have override tags / numbers to override the general Amber Fe parameters - but this must be set up in a way where they don't override each other.

**Comment by jchodera** [comment]
Let's think about the most common user use cases.

The simplest case is that the user just wants to use a prepackaged forcefield set:
```python
forcefield = ForceField('amber14.xml')
```
Now, suppose we want to replace the solvent model with TIP4P-Ew.  We could either just build in the override potential to our solvent and ion models and let them say
```python
forcefield = ForceField('amber14.xml', 'tip4pew.xml')
```
or we could require them to also add the override flag
```python
forcefield = ForceField('amber14.xml', 'tip4pew.xml', allow_overrides=True)
```
though this seems cumbersome for novice users.  

I think it would be easiest on the novice user to let the forcefield developer encode the override order, but I could see it being very useful for power users or people debugging forcefield issues to also add an `allow_overrides` option to either/both `ForceField()` constructor or `ForceField.loadFile()` so that we can override the `priority` index with the number of the file being loaded.

So, I'd argue for both:
* add the `priority` attribute to `<ForceField/>` tag reading, with a default of 0
* add an optional `override` flag to minimally `ForceField.loadFile()` that sets the priority flag to `nfiles_read+1` regardless of the `priority` tag (and possibly also to the constructor)

**Comment by peastman** [comment]
> So, I'd argue for both:

Let's go through that in detail and make sure we're all talking about the same thing.  Is this a correct statement of what you're proposing?

1. If you pass multiple files to the ForceField constructor, the order you list them doesn't matter.  In case of conflicts, it takes the definition with the higher priority.  If they both have the same priority, it throws an exception.

2. If you call `loadFile()` and specify `override=True`, any definitions in that file override previously loaded definitions regardless of their priorities.

3. If you call `loadFile()` and do *not* specify `override=True`, it behaves as if you had simply included that file in the constructor: it will override or not based on priority.

4. What if you call `loadFile()` twice, specifying `override=True` the first time and `override=False` the second time?  Can the second file override the first if it has higher priority?

> Also, if you're going to have a template that overrides others and it can use already existing 'general' atom types, then why complicate this further by adding new atom types with different names but with the same parameters?

No reason at all.  You would only define new atom types if you want them to have *different* parameters.

**Comment by jchodera** [comment]
> What if you call loadFile() twice, specifying override=True the first time and override=False the second time? Can the second file override the first if it has higher priority?

The second call would default to the `priority` specified in the `ffxml` file, which if omitted is 0.

I do think it's important that both residues *and* parameters be allowed to override earlier definitions through this mechanism.

**Comment by peastman** [comment]
> I do think it's important that both residues and parameters be allowed to override earlier definitions through this mechanism.

Could you explain why?  What are the use cases that cannot be addressed by defining new atom types when you want new parameters?  Keep in mind that it's enormously more complicated (both conceptually and in terms of implementation), and leads to a lot of nonobvious corner cases.  (Can an improper torsion definition override another definition that has the same types in a different order?  Does a high priority definition involving wildcards override a low priority one without wildcards?  What do you do when one definition uses types and the other uses classes, such that they cover different but overlapping sets of atoms?  What do we do about user-written generators that don't have any provision for overriding?  etc.)

**Comment by rafwiewiora** [comment]
> Keep in mind that it's enormously more complicated (both conceptually and in terms of implementation), and leads to a lot of nonobvious corner cases. (Can an improper torsion definition override another definition that has the same types in a different order? Does a high priority definition involving wildcards override a low priority one without wildcards? What do you do when one definition uses types and the other uses classes, such that they cover different but overlapping sets of atoms? What do we do about user-written generators that don't have any provision for overriding? etc.)

Good you asked these questions. Because I asked myself - what does the current code do? And it turns out that overwriting parameters ALREADY happens and the order of the ffxml files MATTERS. I'm not sure if I was the only one who didn't realize this, or if we all have had a wrong interpretation of the current situation in this discussion. 

I was very surprised to find that both:
* loading an ffxml that specifies a new set of parameters for the same Atom Type / group of Types
* specifying multiple parameters in the same ffxml

is possible and loads into 'ForceField' with no exception, system can be created.  This is as long as no atom type is specified in `<AtomTypes>` with the same name, because this: https://github.com/pandegroup/openmm/blob/master/wrappers/python/simtk/openmm/app/forcefield.py#L218

``` python
if name in self._atomTypes:
   raise ValueError('Found multiple definitions for atom type: '+name)
``` 

is the ONLY exception that gets thrown if there's something multiple in an ffxml or a sequence of ffxmls. 

Let me illustrate:

PeriodicTorsionForce - propers:
https://github.com/pandegroup/openmm/blob/master/wrappers/python/simtk/openmm/app/forcefield.py#L1350
``` python
if (type2 in types2 and type3 in types3 and type4 in types4 and type1 in types1) or (type2 in types3 and type3 in types2 and type4 in types1 and type1 in types4):
    hasWildcard = (wildcard in (types1, types2, types3, types4))
    if match is None or not hasWildcard: # Prefer specific definitions over ones with wildcards
        match = tordef
    if not hasWildcard:
        break
```

There is no check for multiple instances and clearly the first set specified (that does not use wildcards) will be used. For example, I took the ff99SB forcefield and converted parm99.dat into parm99.xml, frcmod.ff99SB into frcmod_ff99SB.xml. The frcmod specifies 4 new propers:
``` xml
<PeriodicTorsionForce>
  <Proper type1="C" type2="N" type3="CT" type4="C" periodicity1="4" phase1="0.0" k1="0.0" periodicity2="3" phase2="0.0" k2="1.75728" periodicity3="2" phase3="0.0" k3="1.12968" periodicity4="1" phase4="0.0" k4="0.0"/>
  <Proper type1="N" type2="CT" type3="C" type4="N" periodicity1="4" phase1="0.0" k1="0.0" periodicity2="3" phase2="3.14159265359" k2="2.3012" periodicity3="2" phase3="3.14159265359" k3="6.61072" periodicity4="1" phase4="3.14159265359" k4="1.8828"/>
  <Proper type1="CT" type2="CT" type3="N" type4="C" periodicity1="4" phase1="0.0" k1="0.0" periodicity2="3" phase2="0.0" k2="1.6736" periodicity3="2" phase3="0.0" k3="8.368" periodicity4="1" phase4="0.0" k4="8.368"/>
  <Proper type1="CT" type2="CT" type3="C" type4="N" periodicity1="4" phase1="0.0" k1="0.0" periodicity2="3" phase2="0.0" k2="1.6736" periodicity3="2" phase3="0.0" k3="0.8368" periodicity4="1" phase4="0.0" k4="0.8368"/>
 </PeriodicTorsionForce>
```

replacing these from the parm99.dat:
``` xml
<Proper type1="N" type2="CT" type3="C" type4="N" periodicity1="1" phase1="3.14159265359" k1="7.1128" periodicity2="2" phase2="3.14159265359" k2="8.368"/>
<Proper type1="C" type2="N" type3="CT" type4="C" periodicity1="2" phase1="3.14159265359" k1="3.5564" periodicity2="1" phase2="0.0" k2="3.3472"/>
<Proper type1="CT" type2="CT" type3="N" type4="C" periodicity1="4" phase1="3.14159265359" k1="2.092" periodicity2="3" phase2="3.14159265359" k2="0.6276" periodicity3="2" phase3="0.0" k3="0.0" periodicity4="1" phase4="0.0" k4="2.21752"/>
<Proper type1="CT" type2="CT" type3="C" type4="N" periodicity1="4" phase1="0.0" k1="0.4184" periodicity2="2" phase2="0.0" k2="0.29288"/>
```

Now the current code means that:
* ` ff = ForceField('parm99.xml', 'frcmod_ff99SB.xml', 'residues.xml')`
* `ff = ForceField('frcmod_ff99SB.xml', 'parm99.xml', 'residues.xml')`

as long as only one of the parameter files contains `<AtomTypes>` are NOT equivalent and 1st case will use parm's propers, while 2nd frcmod's propers. No exceptions anywhere. Putting both sets into the same file also goes through and again the first set will be used. 

The situation is the same for HarmonicBondForce, HarmonicAngleForce and Impropers - check out the code, they all have the iteration through params - break when satisfied. First listed file currently wins!

For NonbondedForce, it is the LAST occurence that currently wins, because there's no loop with a break in this case and we're just overwriting in a dictionary:
https://github.com/pandegroup/openmm/blob/master/wrappers/python/simtk/openmm/app/forcefield.py#L502
``` python
for t in types[0]:
    self.paramsForType[t] = values
```

BTW this came from my realization finally that the frcmod files are the use case for Amber this is a must for! In our plan to convert every Amber file to an ffxml and provide meta-files with the suggested combinations, if frcmods are to do what they do in Amber, i.e. overwrite some older parameters - we must be able to overwrite parameters.

**Comment by rafwiewiora** [comment]
So with a changed picture of the situation I propose to contain it by normalizing the situation to last one wins, without an argument - unless you want to refactor the code to add Exceptions everywhere for multiple instances and then use an overwrite=True argument? 

We don't even have to do anything to the templates, the last one already wins!

**Comment by peastman** [comment]
In the current system, it is explicitly a bug to have multiple rules that match the same atom/bond/etc.  It's up to force field developers (that's us) to make sure it never happens.  If it ever does happen, the results are undefined, different for different forces, and subject to change.  Arguably we should have error checking code to catch this bug in the force field, but that's a separate question.

Anyway, this is irrelevant to the questions I asked.  Read them again.  You'll see that "last one wins" is not a meaningful answer to them.

**Comment by rafwiewiora** [comment]
> It's up to force field developers (that's us) to make sure it never happens.

Will that change? Can you think of developers setting up ffxmls in any way that will cause the problems you're thinking about in your questions? All we're discussing here is giving mostly the developers a better platform to make ForceFields, with very clear use cases for users. 

> Anyway, this is irrelevant to the questions I asked. Read them again. You'll see that "last one wins" is not a meaningful answer to them

It is relevant in that it is not true that we are introducing a feature that will suddenly cause all of these issues. These issues exist already and it's fine to introduce some rules that will actually standarize the situation more and say that the overriding is only defined for such particular, well-defined cases. Anything else more exotic just stays as it is. 

> Could you explain why? What are the use cases that cannot be addressed by defining new atom types when you want new parameters? 

Use case: frcmod files. I want `ForceField('parm99.xml', 'frcmod_ff99SB.xml')` to override the parameters from the parm with the updated parameters from the frcmod. There is no discussion here - this is what Amber does, it's a clean system, it's much better than making combinations of parm-frcmod's just because we can't override a few parameters. 

And I'm asking for this pretty much as a developer feature, just to enable me to write the master ffxml files - I want to know, that if in that file I will include parm99.dat first, then frcmod_ff99SB.xml, I will get the ff99SB parameters. And this master file will be used by most users. Whether that is done through last one wins, priority tags, special code for dealing with master files - that's secondary, but we can finally agree that parameters must have the ability to override? We will be providing all the component files separately, so there must be a clear outcome of a `ForceField('parm99.xml', 'frcmod_ff99SB.xml')` call. The people who will be using this will most likely be ok with dealing with the fact that order of files matters - last one wins would be great. 

In this use case, the issues you worry about in your questions won't matter - this overriding will be very clean as long as we agree on:

> (Can an improper torsion definition override another definition that has the same types in a different order? Does a high priority definition involving wildcards override a low priority one without wildcards? What do you do when one definition uses types and the other uses classes, such that they cover different but overlapping sets of atoms? What do we do about user-written generators that don't have any provision for overriding? etc.)

* order has to be the same
* no such wildcard business
* no such type/classes mixing, anyway we're not using classes at all in the new conversion
* if they need overriding they'll code it in, if they don't - keep the current behavior. 

I want to strongly stress the separation between the two use cases:

* solvents or ions overriding 'general' definitions - this will be done by users and I agree must be very clear and not mistake prone. Last one wins only is clearly not a good idea here. I'll settle on priority tags, all new atom types, we control everything. Just going back to this:

> No reason at all. You would only define new atom types if you want them to have different parameters.

This works under an assumption that the 'general' FF will always have all the needed atom types. This doesn't have to be true. For my Zn-dummy atom ffxml, I will have new atom types for the zinc cluster, but I also need to make sure that the negative cysteine parameters are available. Do I rely on the assumption that these will be in every Amber forcefield and have to check that every time I introduce a new metal model / a new Amber forcefield? Doesn't sound good. I'll have to include the CYM in the Zn ffxml too, JUST IN CASE - so either I have that also as an overriding template with new atom types - we'll be making quite a few new atom types, but I'm willing to accept that; or I'm allowed to have multiple files with the same atom types / parameters (i.e. if the general FF has CYM, but also my ion file had it just in case.)

* frcmod's - parameter overriding is a must, need clear guidance on what the master ffxml file is to do, API should support a `ForceField` call with overriding too - last one wins is ok for the much more advanced users who will be using this. `overwrite=True` - sure, let's be safer.

**Comment by jchodera** [comment]
> Could you explain why? What are the use cases that cannot be addressed by defining new atom types when you want new parameters? Keep in mind that it's enormously more complicated (both conceptually and in terms of implementation), and leads to a lot of nonobvious corner cases. (Can an improper torsion definition override another definition that has the same types in a different order? Does a high priority definition involving wildcards override a low priority one without wildcards? What do you do when one definition uses types and the other uses classes, such that they cover different but overlapping sets of atoms? What do we do about user-written generators that don't have any provision for overriding? etc.)

The crux of the matter is that other major codes (AMBER and CHARMM) support the capability to override existing parameters, and we have seen instances of its use in our conversion efforts.  What you're asking is "if you do this other thing instead, will that be sufficient?"  The honest answer is that I don't know.  I don't know if it is sufficient for us to get through the current conversion effort for AMBER and CHARMM or how much extra work it adds on our end, and I certainly don't know whether it will ultimately be too limiting.

I think the more relevant question is just "can we punt on this decision for now?"  We can try to forge ahead and see if we run into insurmountable problems with the AMBER and CHARMM conversions.  It sounds like @rafwiewiora already has trouble here, and @ChayaSt would need to hand-curate some forcefields in order to eliminate the parameters that appear more than once after guessing which one is intended.  This kind of manual curation introduces all sorts of issues, and I have big problems with anything that isn't fully reproducible or automated.  Additionally, the existing behavior allowed some degree of (buggy) overriding, so we need to do *something* about it now---this might be a good opportunity to fix the behavior to match what we want.

We certainly can't expect to change the way other codes work or influence how people use them to develop forcefields.  I can't really predict whether the idea of creating new atom types to get around the need to change existing parameter definitions will be sufficient---it's obviously a very different thing.  What I'd like to avoid is a case where we're boxed into having to convert entire forcefields at once into combinations that are only usable together with little additional flexibility, since this will be too limiting.

Let's talk about this more in our chat tomorrow, but I think it might be worth revisiting the concern that users accidentally putting files in the wrong order is a real potential source of user error.  I think everything else we're looking at is just so much more complicated that maybe this is really the simplest route after all.

**Comment by peastman** [comment]
> It is relevant in that it is not true that we are introducing a feature that will suddenly cause all of these issues.

I think you're misunderstanding.  Any behavior that is not documented is undefined.  Any user who relies on undefined behavior is likely to end up getting hurt.  None of the existing force fields rely on undefined behavior.  Therefore, this is not currently an issue.

What we're talking about is a *change* to the documented behavior, thus inviting users to do new things.  Our goal in this discussion is to define what the new documented behavior should be.

Part of what we're discussing is a new feature: an ability to indicate a priority for what files should override what other ones.  That new feature will then interact with lots of other existing behaviors (defined and undefined).  Those interactions could potentially introduce a lot of strange corner cases with behaviors that would be very non-obvious.  That's a bad thing.  I gave a few examples of some of those strange, non-obvious corner cases.  If we allow parameter rules for existing atom types to be overridden, it will necessarily create a *lot* of those strange, non-obvious corner cases.  It will also be impossible for us to ensure consistent behavior, because the handling of overriding would be split between many different pieces of code, some of which are not written by us.  All of those are powerful arguments against allowing that.

Perhaps we'll conclude that's still the best option, that it's the only way to satisfy our use cases.  But to decide that, we need to enumerate all our use cases and consider the possible ways of satisfying each one.

frcmod files are not a use case.  They are a feature of Amber, not OpenMM.  If you want to load Amber files, we already have ways of doing that.  We want to repackage the Amber force fields in a way that lets them be loaded into OpenMM, but that doesn't mean they need to be organized and split into files in precisely the same way they are in Amber.

For example, you mentioned creating the ff99SB force field by starting from ff99, then patching it to modify the torsions.  Well, we already provide the ff99SB force field, and that isn't how we do it.  Instead we merge them into a single file.  I would argue that for end users, our approach is better, and for developers, it's at least as good.  (See my comments at https://github.com/pandegroup/openmm/issues/1344#issuecomment-173685119.)

What I'm looking for are specific examples of things that end users need to do, or that force field developers need to do, that either *cannot* be done without this feature, or are *significantly more difficult* to do without this feature.  That doesn't mean they need to be done in exactly the same way as in other codes.  I've never felt bound to do things in the same way as other codes.  When I think a different way is better, I don't hesitate to use it.  That's precisely why OpenMM has so many unique features that set it apart from other codes!  But there should be *some* reasonable way of doing it.

**Comment by peastman** [comment]
I totally agree that force field conversions should be completely automated.  Anything that requires manual editing of parameters is begging for trouble.  For example, if one Amber file overrides parameters in another one, we need to process all of those overrides and determine the final list of parameters in a completely automated way.  That processing could be done at two different times: when you generate the OpenMM XML files, or when ForceField loads the XML files.  Both are options, but whichever one we end up using, it needs to be completely automated.

It appears to me that Amber force fields use overriding for two very different purposes:

1. Sometimes it simply reflects the historical evolution of force fields.  For example, ff99SB was created by starting from ff99, then changing certain parameters.  That's an interesting historical fact, but not really relevant to how they're used.  They're really independent force fields.  Any simulation will use one or the other, but never do any sort of "mix and match" between them.  Processing the overrides when the force field is converted will therefore work fine.

2. In other cases, it reflects modelling choices where the user might legitimately want to mix and match.  For example, you might choose the ff99SB force field, but then want to combine that with a variety of water models or ion models.  These are the cases where we really need to do the processing within ForceField.

> @ChayaSt would need to hand-curate some forcefields in order to eliminate the parameters that appear more than once after guessing which one is intended.

Could you elaborate on this?  If you specify a particular list of Amber or CHARMM input files, no guessing should ever be required!  It should be completely unambiguous what the final set of parameters are.

**Comment by jchodera** [comment]
> Could you elaborate on this? If you specify a particular list of Amber or CHARMM input files, no guessing should ever be required! It should be completely unambiguous what the final set of parameters are.

Full gory details are here: https://github.com/ParmEd/ParmEd/issues/561

**Comment by swails** [comment]
A recent thread on the Amber mailing list seems to support some of @peastman's arguments.  It hasn't shown up on the archives yet (when it does, it will be here: http://archive.ambermd.org/201602/), so I will first post the initial question until the archive appears, then I'll edit this comment and just provide a link to it:

> Dear Amber users,
> 
> I am simulating a RNA system using the ff14SB in Sander. I am facing an
> issue with  SP3 planar nitrogen atoms of my system. After minimization, the
> orientation of planar SP3 Nitrogen is changing and its planar conformation
> is lost.
> 
> I have checked the same system with older ff99bsc0. This ff works fine and
> the SP3 planar nitrogen atoms are still planar after minimization.
> 
> I am not getting a clue into this...
> 
> can some one give me a solution to this issue????
> 
> Other parameters being loaded with ff14SB are:
> 
> frcmod.modrna08
> frcmod.chiOL4
> frcmod.ionsjc_tip3p
> 
> while in case of ff99bsc0 i am using:
> 
> leaprc.rna.ff99
> leaprc.ff99bsc0
> frcmod.ol.dat
> 
> What is the reason of this??
> 
> Mr. Prayagraj Fandilolu
> Research Scholar,
> Structural Bioinformatics Unit
> Department of Biochemistry
> Shivaji University, Kolhapur - 416004
> (MS) India

Here is my response:

> > Other parameters being loaded with ff14SB are:
> > 
> > frcmod.modrna08
> > frcmod.chiOL4
> 

> ​I'm not sure why you are loading these.  The manual describes what force field each leaprc contains.  Since you seem to be (??) loading extra parameters aside from what is loaded inside each leaprc itself, you are *not* actually using ff14SB here.
> ​ 

> > frcmod.ionsjc_tip3p

> 
> ​Loading ion parameters is necessary, as you've done here.
> ​ 
> 
> 

> while in case of ff99bsc0 i am using:
> > 
> > leaprc.rna.ff99
> > leaprc.ff99bsc0
> > frcmod.ol.dat

> 
> ​Same comments above.  leaprc.ff99bsc0 (now located in oldff/leaprc.ff99bsc0) is the full specification of that force field.
> 
> ​The "extra" parameters you loaded may have been overwriting "good" parameters with bad ones.
> 
> HTH,
> Jason

And the force field documentation seems to me to be pretty clear, and had he read/understood the documentation this person almost certainly would not have tried what he tried... maybe.  The point is that he borked his whole force field doing things that, on the surface, aren't obviously wrong things to do unless you're an Amber proficient.

We have to balance here between simplicity of FF developer's tasks and how easy it is to do something bad/wrong from a user's perspective.

In my opinion, what we need here is improved error checking (since I suspect the balance will have to strike some kind of compromise and will require overwriting capabilities) and some kind of priority-checking that will explicitly allow parameter overrides (but we still want to make sure we're not doing it by accident).

What I *don't* think we want to do is do a 1-to-1 file mapping of Amber or CHARMM files to XML files and have `ForceField` duplicate the logic of the program from which the files were converted.  Because then that inherits all of the problems and complexity of the parent program.

**Comment by jchodera** [comment]
> What I don't think we want to do is do a 1-to-1 file mapping of Amber or CHARMM files to XML files and have ForceField duplicate the logic of the program from which the files were converted. Because then that inherits all of the problems and complexity of the parent program.

We also have to balance this with the difficulty of performing this conversion and the difficulty of anticipating all the possible things that users *can* do with the current system (and should be allowed to do) with what they end up being allowed to do in OpenMM.  If there is a significant mismatch between those capabilities, it will turn off users from OpenMM.

Also, we simply can't do things that will require a lot more effort from our end in the conversion.  That basically means either a 1:1 mapping of files (with some additional recommended bundles) or monolithic conversion (which we disfavor because this is hugely limiting).

**Comment by jchodera** [comment]
One other corner case that hasn't been discussed much: Adding parameter definitions that are *identical* to existing parameter definitions.

When generating small molecule parameters with `antechamber`, `parmchk2` will generate new parameters missing from `gaff.dat`.  The use cases for OpenMM are:
* a user batch-converts a lot of molecules at once into `molecules.xml` that is intended for use with `gaff.xml`
* a user uses a template generator plugin to generate additional residues on the fly
* a user explicitly generates residue templates and additional `ffxml` parameters and calls `ForceField.loadFile()` to load the new parameters

It is possible that when working with multiple molecules (or sets of molecules) `parmchk2` may produce multiple definitions (presumably identical!) for the same parameter.  That might mean that files `moleculeset1.xml` and moleculeset2.xml` have some identical parameters defined.

**Comment by swails** [comment]
> Also, we simply can't do things that will require a lot more effort from our end in the conversion. That basically means either a 1:1 mapping of files (with some additional recommended bundles) or monolithic conversion (which we disfavor because this is hugely limiting).

Well the easiest thing to do on our end is *not* 1:1 mapping of files.  Guaranteed.  IMO, it's not even easiest from yours, mine, and @rafwiewiora's perspective.  1:1 mapping of leaprc files is easiest -- it requires a Python script of literally 2 lines, the first being an import:

```Python
import parmed as pmd
pmd.openmm.OpenMMParameterSet.from_parameterset(pmd.amber.AmberParameterSet.from_leaprc('leaprc')).write('blah.xml')
```

You can't get easier than that.  It's also the most straightforward conceptual mapping for such a conversion.

A 1-to-1 conversion on a file-by-file basis is almost as easy for us *to create the files*, but way harder for us to figure out how to use them correctly and way harder for @peastman to implement in ForceField.

I'm pretty convinced that trying to shoe-horn Amber's implementation of its various force fields exactly into OpenMM's XML machinery is a mistake.  Maybe we should look at GROMACS as more of an example here (whose infrastructure for doing this kind of stuff is much newer than Amber's and has presumably learned from past mistakes better).

**Comment by peastman** [comment]
> Full gory details are here: ParmEd/ParmEd#561

If I'm understanding that discussion correctly, it sounds like the problem is a bug in the CHARMM input files?  If so, that's not something we can fix in OpenMM or ParmEd.  If there's an ambiguity in the force field definition, the best we can hope to do is detect the problem, then get the original force field developers to fix it.

> Also, we simply can't do things that will require a lot more effort from our end in the conversion. That basically means either a 1:1 mapping of files (with some additional recommended bundles) or monolithic conversion (which we disfavor because this is hugely limiting).

Actually, that's *exactly* how I see my own role.  As a software engineer, my job is to do lots of work and solve hard problems so that users of my software don't have to.  Obviously, your position is different: your primarily role is being a scientist, not a software engineer, and time spent converting force fields from one format to another is just a distraction from your main work.  But for me, this is my main work, and I'm willing to put in a fair amount of time to make sure it comes out right.  (A corollary is that I always want to spend time researching each feature and making sure we've found the best design, rather than just doing *something* quickly.  I know that sometimes conflicts with your research priorities, but I do think it's in the best interest of the long term health of the software.)

Besides, I don't think it's as much a dichotomy as what you say.  We can convert *some* groups of files into single XML files, while having other files or groups of files converted separately.  We don't have to *either* do a 1:1 mapping, *or* put absolutely everything into a single file.  We can merge some files but not others.

**Comment by peastman** [comment]
> One other corner case that hasn't been discussed much: Adding parameter definitions that are identical to existing parameter definitions.

I think we can deal with that.  Currently it's a bug to have duplicate definitions.  We don't actually detect that error, as @rafwiewiora noted, but we should.  Assuming we add that error checking code, we could make it be tolerant of multiple identical definitions.  It would only throw an exception if it found multiple definitions that were *not* identical to each other.

**Comment by jchodera** [comment]
> If I'm understanding that discussion correctly, it sounds like the problem is a bug in the CHARMM input files? If so, that's not something we can fix in OpenMM or ParmEd. If there's an ambiguity in the force field definition, the best we can hope to do is detect the problem, then get the original force field developers to fix it.

Yes, the CHARMM developers consider this a bug, *but it doesn't prevent their users from using the files at all*.  They'll fix it in a month or two, but there is no great hurry because it doesn't bring things to a grinding halt.  If our failure mode is "everything comes to a screeching halt and nobody can use CHARMM as a result", that's not quite ideal.  :)

**Comment by swails** [comment]
Sorry for being blunt, but are you seriously suggesting that ForceField's infrastructure adapt to accommodate buggy behavior inside CHARMM (for what is most likely a corner case)?  That seems crazy!  How is that in any way *better* for our users than tweaking our conversion script to work around the bug and implement what we *think* is the intended behavior in the first place?

**Comment by jchodera** [comment]
> Sorry for being blunt, but are you seriously suggesting that ForceField's infrastructure adapt to accommodate buggy behavior inside CHARMM (for what is most likely a corner case)? That seems crazy! 

It certainly isn't crazy.  These are production simulation codes that are used by many many people.  While the duplicated parameter definitions may result in a slightly suboptimal forcefield, it still runs, and according to the implementations in those programs, *that is the canonical CHARMM36 forcefield* for purposes of doing science and reproducing results.  I am absolutely suggesting we should be able to reproduce the science going on in other codes right now.

Just because someone fucked up and screwed up a parameter slightly, that does not mean nobody should be able to use CHARMM in OpenMM.

**Comment by jchodera** [comment]
> How is that in any way better for our users than tweaking our conversion script to work around the bug and implement what we think is the intended behavior in the first place?

We should *never* make that decision, especially when it contradicts what is currently actually done by these codes.  If we want reproducibility, we are not at liberty to decide what was or was not intended by the people who made those files---we only need to ensure we reproduce the correct behavior.

**Comment by jchodera** [comment]
> How is that in any way better for our users than tweaking our conversion script to work around the bug and implement what we think is the intended behavior in the first place?

We could have the conversion scripts try to replicate this behavior, but this is impossible if the parameter redefinitions occur in different files.  That's why our choices come down to (1) allow some way for `ForceField` to handle redefinitions to produce the same behavior as CHARMM, or (2) only use monolithic combinations of forcefields.  I really disfavor (2) because it is highly limiting and inflexible. 

We can discuss more tomorrow maybe?

**Comment by swails** [comment]
> We should never make that decision, especially when it contradicts what is currently actually done by these codes. If we want reproducibility, we are not at liberty to decide what was or was not intended by the people who made those files---we only need to ensure we reproduce the correct behavior.

Allow me to rephrase:  We should work around the bug to implement what CHARMM *actually* does *without* having to compromise the design of our software.  I only said "intends" because, by your own admission *nobody, including the CHARMM people, even know what the actual behavior is.*

Your proposal is to whack on ForceField until it does what CHARMM does for CHARMM files.  My proposal is to whack on the ParmEd conversion script so the resulting ForceField does what CHARMM does for CHARMM files.  Either approach is going to require the same kind of validation.

> Just because someone fucked up and screwed up a parameter slightly, that does not mean nobody should be able to use CHARMM in OpenMM.

People can always use CHARMM in OpenMM -- the same way they do now (via PSFs and parameter files).  CHARMM parameter file conversions will let people use a more OpenMM-like workflow.

**Comment by swails** [comment]
> (2) only use monolithic combinations of forcefields. I really disfavor (2) because it is highly limiting and inflexible.

I actually don't see how this is limiting and unflexible, personally.  GROMACS does it.  OpenMM currently does it for the force fields it supports.  The only knock I've heard on both approaches, from a user's perspective, is the fact that it has no good way of handling small molecules the way that Amber does with antechamber and GAFF.  But that's a completely different problem that is being tackled in parallel to this one.

GROMACS also has a way of *explicitly* overriding some of its parameters, by defining them inside the residue templates themselves (or something along those lines).  This is how ff99SB-ildn is implemented.

You're also missing a third choice -- dodge the collision inside the ParmEd conversion script and write out a set of OpenMM XML files that don't have the error, but still implement the same (buggy) CHARMM force field.  Sure, that's the hardest approach for us, but it's certainly not intractable.

**Comment by swails** [comment]
And yes, we can discuss more tomorrow.

**Comment by jchodera** [comment]
>> (2) only use monolithic combinations of forcefields. I really disfavor
(2) because it is highly limiting and inflexible.
>
> I actually don't see how this is limiting and unflexible, personally.

I don't see how we could possibly anticipate all combinations of
forcefields people would want to use. You have seen the CHARMM parameter
file directory, yes? There are a huge variety of possible permutations
someone might want to use. And for Amber, do you always want to use the
Glycam FF? Should we make every possible combinations of Glycam with every
other compatible ForceField? And every combination of solvents or ions? We
definitely need some sort of override mechanism. The only major question is
"which one?".

> You're also missing a third choice -- dodge the collision inside the
ParmEd conversion script and write out a set of OpenMM XML files that don't
have the error, but still implement the same (buggy) CHARMM force field.
Sure, that's the hardest approach for us, but it's certainly not
intractable.

Since the duplicated parameters span multiple files, this requires either
(1) an override mechanism, (2) manual curation to eliminate duplicates, or
(3) a monolithic ffxml where all ForceField files are loaded and processed
and a single monolithic ffxml is spit out by parmed.

I honestly have yet to hear a good argument for why supporting the "last
one wins" behavior is suboptimal. It is actually very powerful for defining
hierarchically typed forcefields and is behavior our new small molecule
forcefield will rely on. The user concerns you have raised can be easily
dealt with by issuing a warning of what is being replaced and why when a
duplicate is encountered, with a flag added to suppress warnings that is
off by default. It is simple, familiar, and mostly already implemented. It
is not a design flaw, but a feature.

**Comment by swails** [comment]
> I don't see how we could possibly anticipate all combinations of
> forcefields people would want to use. You have seen the CHARMM parameter
> file directory, yes? There are a huge variety of possible permutations
> someone might want to use. And for Amber, do you always want to use the
> Glycam FF? Should we make every possible combinations of Glycam with every
> other compatible ForceField? And every combination of solvents or ions? We
> definitely need some sort of override mechanism. The only major question is
> "which one?".

I think you're misinterpreting what I'm suggesting (which I think is similar to what @peastman has suggested).  What I see happening is convert each leaprc to a single XML.  They are each separate force fields and billed as such.

As they are declared compatible, you can mix-and-match them in the same way that Amber does.  I'm referring to ff14SB.xml, GLYCAM.xml, and lipid14.xml when I say "monolithic".  I view the leaprc files as the atomic units in Amber force fields, *not* the parm.dat/frcmod files.  The only use-case that users have for frcmod files is for modified residues.  Which means that nobody ever loads frcmod.ff14SB unless they've made a mistake.  So to me, converting frcmod.ff14SB into frcmod_ff14SB.xml makes no sense.  FF developers also use frcmods, but FF development in OpenMM need not be implemented exactly the same way as it is in Amber.  It's quite easy to effect the same thing with an explicity priority-based approach.

And I can speak from the experience of multiple conversations and discussions with the authors of each of those force fields that the intended behavior of combining those force fields is that *no parameter ever gets overwritten*.  If it does, it is a bug.  And moreover, it's a bug with *no well-defined behavior*.  Combining lipid14 and ff14SB (and GLYCAM) has no defined order.  Users are on their own to source each leaprc in whatever order they want.  So if Amber does accidentally use overlapping types, "last one wins" is completely arbitrary.  The fact that Amber takes a "last one wins" approach to storing a parameter database inside tleap means Amber devs can implement leaprc files as an underlying force field with a series of frcmod modifications, but it also means that it is impossible to detect when FFs we claim to be compatible actually have clashes.  This is where "feature" becomes "flaw".

My understanding of the CHARMM force field is far less complete, but I believe they have a similar issue with users defining combination orders.  Sure, a specific input file gives well-defined behavior, but the fact that each person has their own input file means that there is no way to implement "The CHARMM Force Field" (TM), as it's used in the wild when these buggy type collisions exist.

> I honestly have yet to hear a good argument for why supporting the "last
> one wins" behavior is suboptimal.

If CHARMM did not support this (do you even *know* if CHARMM supports this? I thought we were all just *assuming* they did but not even the CHARMM folks could confirm this), *this bug would not exist.*  How is that not a good argument? The only place we're sure "last one wins" is implemented is in Amber, and that "feature" is used exclusively inside leaprc files from a user's perspective.  If leaprc files get translated 1-to-1 to XML files, this behavior moves from "feature" to "bug".  Now for my part, I'm not suggesting there are no good reasons to support "last one wins" -- there are legitimate arguments.  But it should be weighed against the pros of *not* doing that.  What I'm looking for is a way to have our cake and eat it too.  Can we find another approach that has both benefits while minimizing drawbacks?  You seem to have already decided that the answer is no (which establishes a barrier to constructive brainstorming IMO).

For my case, I have yet to hear a good argument for why, at least for Amber, file conversions should be done at the level inside dat/leap/parm (where the parm.dat/frcmod files live) rather than dat/leap/cmd (where the leaprc files live).

Also, you may have mentioned this before, but could you describe a specific use-case where simply converting all the c36 files into a single XML is *less* flexible than splitting them into separate ones? (I'll note that this does *not* preclude you from combining it with other XML files should the need arise.)  Because from a FF conversion perspective, it's no harder to do one than the other.  And from a user's perspective, monolithic FFs are a bit simpler to use.

**Comment by jchodera** [comment]
> For my case, I have yet to hear a good argument for why, at least for Amber, file conversions should be done at the level inside dat/leap/parm (where the parm.dat/frcmod files live) rather than dat/leap/cmd (where the leaprc files live).

The answer is very simple:
https://github.com/choderalab/ambermini/blob/master/share/amber/dat/leap/cmd/leaprc.ff14SB#L107-L113
```
#
#       Load water and ions
# 
loadOff atomic_ions.lib
loadOff solvents.lib
HOH = TP3
WAT = TP3
```
If you want to use something else besides `atomic_ions.lib` and `solvents.lib`, you're out of luck with this mechanism.

TIP3P is awful.  And the single-site ions won't work for half the PDB, which requires multisite ion models.

So at the *very least* we need a mechanism to override solvents and ions.

**Comment by jchodera** [comment]
> (do you even know if CHARMM supports this? I thought we were all just assuming they did but not even the CHARMM folks could confirm this)

Several parts of the CHARMM documentation state this is how the parsing operates, [for example](http://www.charmm.org/documentation/c35b1/io.html):
```
        NBFIXes are processed in order. For that reason, wildcard
usage should come first. In the case of duplicate specifications,
there is no check, and the last specification will be used.
```

**Comment by jchodera** [comment]
> If CHARMM did not support this (do you even know if CHARMM supports this? I thought we were all just assuming they did but not even the CHARMM folks could confirm this), this bug would not exist

I don't need to point out again that this is [how forcefields like ff99SB are defined](https://github.com/choderalab/ambermini/blob/master/share/amber/dat/leap/parm/frcmod.ff99SB)?  By overriding previous parameters?  Why do you not consider this a bug in AMBER as well?

**Comment by swails** [comment]
> So at the very least we need a mechanism to override solvents and ions.

Easily done by using priorities -- just give the water_ions.xml file for each water model a higher priority than the underlying base force field.  This is an argument for the water/ion .dat files being given separate XML files (as is currently done in OpenMM), but that's already how Amber handles things.  And Amber really doesn't go far enough -- it should provide water_ion leaprc files with the correct combinations.

> I don't need to point out again that this is how forcefields like ff99SB are defined? By overriding previous parameters?

Who cares if we just translate leaprc files?  Why are you so insistent that "ff14SB.xml" be *implemented* exactly the same way as leaprc.ff14SB (i.e., translating parm10.dat into an XML file, frcmod.ff14SB as an XML file, and then having a master XML file use xml directives to stitch them together)?  Where is the use-case for using frcmod.ff14SB outside of the context of the whole force field?  Same for parm10.dat?  Every case of doing that in Amber is a bug.  For example, you read the ff14SB paper and see that frcmod.ff14SB contains the modifications to ff99SB.  So you would expect to be able (in Amber) to do something like this:

```
source leaprc.ff99SB
loadAmberParams frcmod.ff14SB
```

And get FF14SB.  tleap wouldn't complain, and you'd get files that seemed quite reasonable.  Only problem is *this is not ff14SB*.  frcmod.ff14SB sits on top of the parm10.dat implementation of ff99SB, not the parm99.dat+frcmod.ff99SB implementation of it (which is what leaprc.ff99SB uses).

> Why do you not consider this a bug in AMBER as well?

My last message was long -- I'll copy just the relevant section:

> The fact that Amber takes a "last one wins" approach to storing a parameter database inside tleap means Amber devs can implement leaprc files as an underlying force field with a series of frcmod modifications, but it also means that it is impossible to detect when FFs we claim to be compatible actually have clashes. This is where "feature" becomes "flaw".

I call it both a feature *and* a flaw, in that it allows FFs to be implemented as a series of frcmod files, but also masks real bugs.

**Comment by peastman** [comment]
This was implemented in #1373.
